execute/dispatcher.go

```diff
@@ -107,14 +107,24 @@ func (d *poolDispatcher) setErr(err error) {
 
 //Stop the dispatcher.
 func (d *poolDispatcher) Stop() error {
+	// Check if we are already closed and return the error if so.
 	d.mu.Lock()
-	defer d.mu.Unlock()
 	if d.closed {
+		d.mu.Unlock()
 		return d.err
```

Comment:
why do we not need to have the lock when reading the error here when we do below?

Reply:
The `setErr` call can set the error in a different goroutine which means the read needs to be protected too.

Reply:
This code does not have read protection. It releases the lock before reading the error, should the code be:

```
if d.closed {
   err := d.err
   d.mu.Unlock()
   return err
}
```

Reply:
Oh that didn't occur to me. It should just be `defer d.mu.Unlock()`. Your code works too.